### ISSUE TITLE
Use uber-go/atomic in place of sync/atomic

### ISSUE THREAD

**Issue by codesome** [prometheus/prometheus#7647]
1. Avoids bugs like https://github.com/prometheus/prometheus/pull/7646 where the atomic variable is not accessed atomically everywhere.
2. Eliminates the need of keeping atomic variables at the top of the struct

Thanks @pstibrany for pointing it out!

**Comment by johncming** [comment]
The case used `sync/atomic` in `tsdb` pkg.

tsdb/chunks/head_chunks.go
```
# fields in struct
ChunkDiskMapper.curFileNumBytes
```

tsdb/head.go
```
# fields in struct
Head.minTime
Head.maxTime
Head.minValidTime
Head.numSeries
Head.lastSeriesID

# variable in function
unknownRefs in Head.loadWAL method
```

**Comment by bwplotka** [comment]
Nice! 

https://github.com/uber-go/atomic

**Comment by jvrplmlmn** [comment]
I would like to take on this issue. It seems something approachable to get familiar with the codebase and start contributing to the project 😃

**Comment by codesome** [comment]
@jvrplmlmn Sure! Go ahead

**Comment by jvrplmlmn** [comment]
First PR that addresses the usage of `sync/atomic` in the `tsdb` package: https://github.com/prometheus/prometheus/pull/7659 

If the goal is to replace it in the whole codebase let me know and I will send following PRs to that one.

**Comment by codesome** [comment]
These are the remaining places after the first PR
```
➜  prometheus git:(master) grep -r --exclude-dir=vendor "sync/atomic"
notifier/notifier.go:	"sync/atomic"
notifier/notifier_test.go:	"sync/atomic"
storage/remote/intern_test.go:	"sync/atomic"
storage/remote/intern.go:	"sync/atomic"
storage/remote/intern.go:	// See https://golang.org/pkg/sync/atomic/#pkg-note-BUG for more info.
storage/remote/queue_manager_test.go:	"sync/atomic"
storage/remote/ewma.go:	"sync/atomic"
storage/remote/ewma.go:	// See https://golang.org/pkg/sync/atomic/#pkg-note-BUG for more info.
storage/remote/queue_manager.go:	"sync/atomic"
storage/remote/queue_manager.go:	// https://golang.org/pkg/sync/atomic/#pkg-note-BUG
web/web.go:	"sync/atomic"
cmd/prometheus/main.go:	"sync/atomic"
tsdb/head_bench_test.go:	"sync/atomic"
```

**Comment by codesome** [comment]
> If the goal is to replace it in the whole codebase let me know and I will send following PRs to that one.

@jvrplmlmn yes, it's for the entire repo. It would be awesome if you could send PRs for the remaining!

**Comment by jvrplmlmn** [comment]
Sure, I will send the pull requests and also add a mechanism to detect on CI that we are not using directly `sync/atomic`.

**Comment by jvrplmlmn** [comment]
Follow up is available at https://github.com/prometheus/prometheus/pull/7683 for review.

**Comment by codesome** [comment]
Closed in #7659 and #7683

**Comment by roidelapluie** [comment]
I am reopening this so we can tackle the CI part.

@jvrplmlmn I have reviewed your inital script. I think a script is not needed, all we need to do is something like:

[[ "$(go list -f '{{ range .Imports }}{{ if eq . "sync/atomic" }}.{{ end }}{{ end }}' ./...)" == "" ]]

**Comment by roidelapluie** [comment]
Is there anyone willing to PR faillint? that sounds great.

**Comment by jvrplmlmn** [comment]
I am cooking a PR :+1:
